### ISSUE TITLE
Declaring React-Router V4 routes within the render() of a Redux Container doesn't work

### ISSUE THREAD

**Issue by theirishpenguin** [ReactTraining/react-router#4171]
I was trying to declare React-Router V4 routes within the render() of a Redux Container doesn't work. If I remove the connect() call in the container clicking on the links works. But if I use connect() then it doesn't.

 **Is there a compatibility issue between RR4 and redux in this area? Or have I just missed something?**

P.S. I've also tried to move the container up the component tree (ie. separate it out in it's own class) but that didn't work either. The only way I could get it to work is if I moved the redux container down the component tree, underneath the component in which I declare the  React-Router V4 routes.

Here is an example of the original container which exhibits the problem...

```

class MessageBoard extends Component {

  return (                                                                                                            
        <div>                                                                                  
          <Match exactly pattern="/" render={ () => (<Redirect to="/games/1"/>)} />
          <Match exactly pattern="/games/:id" component={GameComponent} />  
          <Match exactly pattern="/super-games" component={SuperGameList} />
       </div>
  )

// I have definitions of redux's mapStateToProps() here - omitted for brevity 

export default MessageBoard // Works
//export default connect(mapStateToProps)(MessageBoard) // Does not work if I try this instead :(
```

This links themselves are defined in another component and each one looks like this...

```
 <Link to={target} activeOnlyWhenExact>{text}</Link>
```

Here is the bug template filled in as best I can...

## Version
react router 4.0.0-2

## Steps to reproduce
As above. If this is deemed a potential issue I can try to put together a fuller example.

## Expected Behavior
Able to declare routes in a redux container. Links works when I click on them

## Actual Behavior
Cannot declare routes in a redux container. Links don't work when I click on them.

**Comment by hopglascock** [comment]
Redux Containers work for me with RR 4.0.0-alpha.5. Perhaps try updating?

**Comment by theirishpenguin** [comment]
Okay, I'll try that or come up a better example. Thanks for the info.

**Comment by theirishpenguin** [comment]
Thanks @hopglascock. I've upgraded to  4.0.0-alpha.5 but I get a bunch of errors...

```
ERROR in ./~/react-router/~/history/createHashHistory.js
Module not found: Error: Cannot resolve module 'warning' in /path/to/project/node_modules/react-router/node_modules/history
 @ ./~/react-router/~/history/createHashHistory.js 7:15-33

ERROR in ./~/react-router/~/history/createBrowserHistory.js
Module not found: Error: Cannot resolve module 'warning' in /path/to/project/node_modules/react-router/node_modules/history
 @ ./~/react-router/~/history/createBrowserHistory.js 9:15-33

ERROR in ./~/react-router/~/history/createMemoryHistory.js
Module not found: Error: Cannot resolve module 'warning' in /path/to/project/node_modules/react-router/node_modules/history
 @ ./~/react-router/~/history/createMemoryHistory.js 9:15-33

ERROR in ./~/react-router/~/history/createHashHistory.js
Module not found: Error: Cannot resolve module 'invariant' in /path/to/project/node_modules/react-router/node_modules/history
 @ ./~/react-router/~/history/createHashHistory.js 11:17-37

ERROR in ./~/react-router/~/history/createBrowserHistory.js
Module not found: Error: Cannot resolve module 'invariant' in /path/to/project/node_modules/react-router/node_modules/history
 @ ./~/react-router/~/history/createBrowserHistory.js 13:17-37

ERROR in ./~/react-router/~/history/LocationUtils.js
Module not found: Error: Cannot resolve module 'resolve-pathname' in /path/to/project/node_modules/react-router/node_modules/history
 @ ./~/react-router/~/history/LocationUtils.js 10:23-50

ERROR in ./~/react-router/~/history/createTransitionManager.js
Module not found: Error: Cannot resolve module 'warning' in /path/to/project/node_modules/react-router/node_modules/history
 @ ./~/react-router/~/history/createTransitionManager.js 5:15-33
```

I'm using the HashRouter at the moment...

```
import React from 'react'                                                       
import { HashRouter } from 'react-router'                                       
import MessageBoard from './MessageBoard' // Our custom react component                         
                                                                                
export const Router = () => (                                                   
  <HashRouter>                                                                  
    <div>                                                                       
      <MessageBoard />                                                                  
    </div>                                                                      
  </HashRouter>                                                                 
)
```

... but I don't think I'm doing anything weird with react-router and the project is a pretty standard one (it's using the latest React and Material-UI). So I'm not sure why the alpha would break like this.

No worries if this is turning into a bit of a rabbit hole. I can work around this for now as per above (it's not a complete workaround as it makes it impossible for me, AFAIK, to get data from a redux store into a top-level component). I can try to update RR4 again when the next version of react-router is released. It's fine by me to close this bug as I haven't provided a very good test case yet.

If I get some time I'll try to put together a test case, but even though the project is straightforward it's big enough to be time consuming to get a reasonable test case.

Thanks for you help.

**Comment by timdorr** [comment]
Try deleting your node_modules and reinstalling. Some important modules aren't getting installed for whatever reason.

**Comment by theirishpenguin** [comment]
@timdorr Yes, I tried that. Same error. Thanks for the suggestion.

**Comment by theirishpenguin** [comment]
Just an update, I retried tackling this problem and it appears that updating from 4.0.0-2 to 4.0.0-alpha.6 fixed this for me.

I'm not 100% sure that this made the difference, however I am pretty sure it did.

As part of the update, I deleted my node_modules directory, deleted my npm shrinkwrap file and had to move from react-tap-event-plugin to 1.0.0 to 2.0.0 in package.json file, as per this answer http://stackoverflow.com/a/40747646/427805. The react-tap-event-plugin version bump was needed as I was getting the error "ERROR in ./~/react-tap-event-plugin/src/injectTapEventPlugin.js" - no doubt because I was updating React as well as React Router).

Anyway, thanks again @timdorr @hopglascock.  I suspect the reason I was getting the "createHashHistory.js" errors above was because I forgot to delete my npm shrinkwrap file at the time. Just a guess.
